libs/wingc/src/lsp/code_actions.rs

```diff
@@ -0,0 +1,238 @@
+		for diagnostic in context.diagnostics {
+			if let Some(mut action) = get_fix_for_diagnostic(uri.clone(), diagnostic) {
+				if let CodeActionOrCommand::CodeAction(action) = &mut action {
+					if action.kind == Some(CodeActionKind::SOURCE_FIX_ALL) {
+						action.kind = Some(CodeActionKind::QUICKFIX);
+					}
+				}
+
+				action_list.push(action);
+			}
+		}
+		return action_list;
+	}
+
+	let mut auto_insert_semicolons = vec![];
+
+	let file_id = uri.to_file_path().unwrap().to_str().unwrap().to_string();
+	let original_diagnostics = get_diagnostics();
+	for original_diagnostic in &original_diagnostics {
+		let Some(original_span) = &original_diagnostic.span else {
+			continue;
+		};
+		if original_span.file_id != file_id {
+			continue;
+		}
+
+		let action = get_fix_for_diagnostic(
+			uri.clone(),
+			Diagnostic {
+				message: original_diagnostic.message.clone(),
+				range: original_diagnostic.span.clone().unwrap_or_default().into(),
+				severity: Some(DiagnosticSeverity::ERROR),
+				..Default::default()
+			},
+		);
+		if let Some(mut action) = action {
+			if let CodeActionOrCommand::CodeAction(action) = &mut action {
+				if action.title == "Insert ';'" {
+					let mut add_autofix = true;
+					// ensure that this error doesn't intersect any other errors
+					for other_diagnostic in &original_diagnostics {
+						if other_diagnostic.message != original_diagnostic.message
+							&& other_diagnostic
+								.span
+								.as_ref()
+								.unwrap()
+								.contains_span(&original_diagnostic.span.as_ref().unwrap())
+						{
+							add_autofix = false;
+							break;
+						}
+					}
+					if add_autofix {
+						auto_insert_semicolons.push((
+							action.diagnostics.as_ref().unwrap()[0].clone(),
+							action.edit.as_ref().unwrap().changes.as_ref().unwrap().clone(),
+						));
+					}
+				}
+			}
+
+			action_list.push(action);
+		}
+	}
+
+	// if there are semicolons missing, add a fix all action to correct them all at once
+	if !auto_insert_semicolons.is_empty() {
+		let mut combined_changes = HashMap::new();
+		for changes in &auto_insert_semicolons {
+			for (file, edits) in &changes.1 {
+				combined_changes
+					.entry(file.clone())
+					.or_insert_with(|| vec![])
+					.extend(edits.clone());
+			}
+		}
+
+		action_list.push(CodeActionOrCommand::CodeAction(CodeAction {
+			title: "Insert ';'".to_string(),
+			kind: Some(CodeActionKind::SOURCE_FIX_ALL),
+			diagnostics: Some(
+				auto_insert_semicolons
+					.iter()
+					.map(|(diagnostic, _)| diagnostic)
+					.cloned()
+					.collect(),
+			),
+			edit: Some(WorkspaceEdit {
+				changes: Some(combined_changes),
+				..Default::default()
+			}),
+
+			is_preferred: Some(true),
+			..Default::default()
+		}));
+	}
+
+	action_list
+}
+
+fn get_fix_for_diagnostic(file: Url, diagnostic: Diagnostic) -> Option<CodeActionOrCommand> {
+	match diagnostic.message.as_str() {
+		"Expected ';'" => {
```

Comment:
Seems fragile... maybe put into a const and reuse this const when the diagnostic is created.

Reply:
Can do. Ideally I'd like the quickfix generation to actually be part of generating the diagnostic so it's not separate, but figured that can wait